File: cpp/answers/3_bfs_answer.cpp
#include <deque>
#include <iostream>
#include <string>
#include <utility>
#include <vector>
using namespace std;

int main() {
    int H, W;
    cin >> H >> W;

    int start_row, start_col;
    int goal_row, goal_col;
    cin >> start_row >> start_col;
    cin >> goal_row >> goal_col;

    vector<string> maze(H);
    for (int i = 0; i < H; i++) {
        cin >> maze[i];
    }

    // 各マスまでの距離。-1 は「まだ訪れていない」
    vector<vector<int>> dist(H, vector<int>(W, -1));

    int dr[4] = {-1, 1, 0, 0};
    int dc[4] = {0, 0, -1, 1};

    deque<pair<int, int>> que;

    // スタート地点をキューに入れる
    dist[start_row][start_col] = 0;
    que.push_back({start_row, start_col});

    while (!que.empty()) {
        auto [row, col] = que.front();
        que.pop_front();

        for (int i = 0; i < 4; i++) {
            int next_row = row + dr[i];
            int next_col = col + dc[i];

            if (next_row < 0 || next_row >= H ||
                next_col < 0 || next_col >= W) {
                continue;
            }

            if (maze[next_row][next_col] == '#') {
                continue;
            }

            if (dist[next_row][next_col] != -1) {
                continue;
            }

            dist[next_row][next_col] = dist[row][col] + 1;
            que.push_back({next_row, next_col});
        }
    }

    int answer = dist[goal_row][goal_col];

    if (answer == -1) {
        cout << "ゴールには到達できません\n";
    } else {
        cout << "最短距離: " << answer << '\n';
    }

    return 0;
}

File: python/answers/3_bfs_answer.py
import sys
from collections import deque

lines = sys.stdin.read().splitlines()
H, W = map(int, lines[0].split())
sy, sx = map(int, lines[1].split())
gy, gx = map(int, lines[2].split())
maze = lines[3:3+H]

# 各マスまでの距離。-1 は「まだ訪れていない」
dist = [[-1] * W for _ in range(H)]

dy = [-1, 1, 0, 0]
dx = [0, 0, -1, 1]

que = deque()

# スタート地点をキューに入れる
dist[sy][sx] = 0
que.append((sy, sx))

while que:
    y, x = que.popleft()

    for i in range(4):
        ny = y + dy[i]
        nx = x + dx[i]

        if not (0 <= ny < H and 0 <= nx < W):
            continue

        if maze[ny][nx] == "#":
            continue

        if dist[ny][nx] != -1:
            continue

        dist[ny][nx] = dist[y][x] + 1
        que.append((ny, nx))

answer = dist[gy][gx]

if answer == -1:
    print("ゴールには到達できません")
else:
    print("最短距離:", answer)
